ext/crates/fp/src/field/smallfq.rs

```diff
@@ -0,0 +1,259 @@
+use std::sync::{Arc, LazyLock};
+
+use dashmap::DashMap as HashMap;
+
+use super::{
+    element::{FieldElement, FieldElementContainer},
+    field_internal::FieldInternal,
+    Field, Fp,
+};
+use crate::{
+    constants::BITS_PER_LIMB,
+    limb::Limb,
+    prime::{log2, Prime, ValidPrime},
+    vector::inner::FqVector,
+    PRIME_TO_INDEX_MAP,
+};
+
+static SMALL_CONWAY_POLYS: [[[u32; 17]; 15]; 54] = include!("small_conway_polys.txt");
+
+type ZechTable = HashMap<SmallFqElement, SmallFqElement>;
```

Comment:
Not quite just an exponent because 0 doesn't belong to the group of units, so rather an
`Option<u32>`. We could still make it work with a Vec by storing 1 at index 0 and $a^n + 1$ at index
$n + 1$. I'm not sure it's worth the mental load, since Hashmap lookup is already O(1). Maybe we
could have a faster custom hash as a compromise?

Reply:
But hash map lookup is a slow O(1) operation whereas array lookup is a very fast O(1) operation. I think it's absolutely worth the mental load, try profiling a bunch of vector additions I'm pretty sure see the time spent hashing show up quite noticeably -- though it might not be as slow as packing and unpacking.

Reply:
I see what you mean now. We're only ever using the Zech table for addition, and therefore only for units. So we can just skip 0 and store the exponents directly. This does have a dramatic effect